resolver_test.go

```diff
@@ -259,6 +259,80 @@ func TestResolverRun(t *testing.T) {
 			t.Fatal("Wrong contents:", string(r.Contents))
 		}
 	})
+
+	// Test issue where Run/Wait would loop forever as each Run would set the
+	// buffering timer, triggering a bufferTrigger which exited Wait..
+	// This shouldn't ever be an issue again with the new implementation.
+	t.Run("buffered-resolve-loop-hang-check", func(t *testing.T) {
+		rv := NewResolver()
+		w := blindWatcher()
+		defer w.Stop()
+		tt := echoTemplate("foo")
+		w.Register(tt)
+		w.SetBufferPeriod(time.Millisecond, time.Second, tt.ID())
+
+		ctx, cancel := context.WithTimeout(context.Background(),
+			time.Millisecond*30)
+		defer cancel()
+
+		// should stop on second loop
+		for i := 0; i < 4; i++ {
+			_, err := rv.Run(tt, w)
+			if err != nil {
+				t.Fatal("Run() error:", err)
+			}
+			select {
+			case <-w.WaitCh(ctx):
+			case <-ctx.Done():
+				return
+			}
+		}
+		t.Error("bad behavior")
+	})
+
+	// Tests that the buffer actually buffers. Uses
```

Comment:
Maybe related - I tried playing around with the min buffer_period. I'm not sure I follow how this test confirms that the buffer waits the buffer time. Would mind elaborating?

Reply:
Forgot to finish that comment.. "Uses" was a start of the description of the fake dependency it uses and how that is used in the test. Let me add that and maybe it will answer your second question (if not, it should and we can fix that).

Reply:
Thanks for adding all of the comments! Really helpful and clarified.